File: blackjack/blackjack.py
from random import choice

suits = ['H', 'S', 'C', 'D'];
numbers = ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K'];


def draw(player1, player2):
    while True:
        card = choice(numbers) + choice(suits);
        if not (card in player1 or card in player2):
            return card


def value(card):
    number = card[:-1];
    if number in ['J', 'Q', 'K']:
        return 10
    elif number == 'A':
        return 1
    else:
        return int(number)


def score(player):
    current_score = 0
    ace = False
    for card in player:
        current_score += value(card)
        if card[0] == 'A':
            ace = True
    if (ace) and (current_score < 12):
        current_score += 10
    return current_score


def hand(player):
    result = player[0]
    for card in player[1:]:
        result += ' ' + card
    return result


first_time = True
while True:
    if first_time:
        first_time = False
    else:
        input("\nPress enter to play again.")
        print("============================")
    dealer = [];
    player = [];

    dealer.append(draw(player, dealer))
    dealer.append(draw(player, dealer))
    player.append(draw(player, dealer))
    player.append(draw(player, dealer))

    if score(dealer) == 21:
        if score(player) == 21:
            print(" Dealer's hand: ", hand(dealer))
            print("     Your hand: ", hand(player))
            print("It's a tie!")
        else:
            print(" Dealer's hand: ", hand(dealer))
            print("     Your hand: ", hand(player))
            print("Dealer Blackjack - You lose...")
    elif score(player) == 21:
        print(" Dealer's hand: ", hand(dealer))
        print("     Your hand: ", hand(player))
        print("BLACKJACK - You win!")
    else:
        print(" Dealer's hand: ", dealer[0], '??')
        print("     Your hand: ", hand(player))

        # Player choices
        while True:
            while True:
                print("What would you like to do?")
                print("0: Stand")
                print("1: Hit")
                decision = input()
                if decision in ['0','1']:
                    break
                print("Invalid input!")
            if decision == '0':
                break
            player.append(draw(player, dealer))
            print(" Dealer's hand: ", dealer[0], '??')
            print("     Your hand: ", hand(player))
            if score(player) == 21:
                break
            elif score(player) > 21:
                print("Bust - You lose...")
                break
            # elif 17 <= score(dealer) < score(player):
            #     print("You win!")
            #     break
        if score(player) > 21:
            continue

        # Dealer choices
        while True:
            print(" Dealer's hand: ", hand(dealer))
            print("     Your hand: ", hand(player))
            if score(dealer) > 21:
                print("You win!")
                break
            elif score(dealer) > score(player):
                print("You lose...")
                break
            elif score(dealer) > 16:
                if score(dealer) < score(player):
                    print("You win!")
                    break
                else:
                    print("It's a tie!")
                    break
            else:
                print("Dealer hits...")
                dealer.append(draw(player, dealer))

File: blackjack/blackjack.js
const suits = ['H', 'S', 'C', 'D'];
const numbers = ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K'];

const hitstick = "<span onclick='hit();'>Hit</span> / <span onclick='stick();'>Stick</span>"
const restart = "<span onclick='init();'>Play again?</span>"
const ask = "Would you like to stick or hit?"


function draw(player) {
  while (true) {
    card = numbers[Math.floor(13*Math.random())] + suits[Math.floor(4*Math.random())]
    if ($("#"+card).length == 0) {
      $("#"+player).append("<img src='cards/" + card + ".png' class='card' id='" + card + "'>");
      return card
    }
  }
}


function hideCard(id) {
  $("#"+id).attr("src", "cards/back.png");
}


function showCard(id) {
  $("#"+id).attr("src", "cards/" + id + ".png");
}


function value(card) {
    number = card.slice(0,-1);
    if (['J', 'Q', 'K'].includes(number)) {
        return 10
    } else if (number == 'A') {
        return 1
    } else {
        return parseInt(number)
    }
}


function score(player) {
  hand = document.getElementById(player).getElementsByClassName("card");
  var currentScore = 0;
  var ace = false;
  for (let i=0;i<hand.length;i++) {
    var card = hand[i].getAttribute("id")
    currentScore += value(card);
    if (card[0] == 'A') {
      ace = true
    }
  }
  if (ace && currentScore < 12) {
    currentScore += 10
  }
  return currentScore
}


function checkStart() {
  if (score("dealer") == 21 && score("player") == 21) {
    // Double blackjack
    showCard(hiddenCard)
    $("#instructions").html("Push...")
    $("#actions").html(restart)
  } else if (score("dealer") == 21) {
    // Dealer blackjack
    showCard(hiddenCard)
    $("#instructions").html("Dealer Blackjack!")
    $("#actions").html(restart)
  } else if (score("player") == 21) {
    // Player blackjack
    showCard(hiddenCard)
    $("#instructions").html("BLACKJACK!")
    $("#actions").html(restart)
  } else {
    // Normal play
    $("#instructions").html(ask)
  }
}


function hit() {
  draw('player');
  if (score('player') > 21) {
    showCard(hiddenCard)
    $("#instructions").html("Bust!")
    $("#actions").html(restart)
  } else if (score('player') == 21) {
    // Dealer plays
    stick()
  }
}


function stick() {
  $("#actions").html(restart)
  $("#instructions").html(" ")
  showCard(hiddenCard)
  dealerDraw()
}


function dealerDraw() {
  if (score('dealer') > 17) {
    if (score('dealer') < score('player') || score('dealer') > 21) {
      $("#instructions").html("You win!")
    } else if (score('dealer') == score('player')) {
      $("#instructions").html("Push...")
    } else {
      $("#instructions").html("You lose...")
    }
  } else {
    setTimeout(function() {
      draw('dealer')
      dealerDraw()
    }, 1000)
  }
}


function reset() {
  var id = window.setTimeout(function() {}, 0);
  console.log(id);
  while (id--) {
      window.clearTimeout(id); // will do nothing if no timeout with id is present
  }
}


function init() {
  reset()
  $("#dealer").empty();
  $("#player").empty();

  $("#actions").html(hitstick)
  draw("dealer")
  hiddenCard = draw("dealer")
  draw("player")
  draw("player")
  hideCard(hiddenCard)

  checkStart()
}
